smelli/unbinned.py

```diff
@@ -0,0 +1,111 @@
+from flavio.io import instanceio as iio
+import numpy as np
+import yaml
+from .util import get_cachepath
+from requests import get
+import os
+from wilson import Wilson
+from sklearn.neighbors import KernelDensity
+import wcxf
+
+class UnbinnedParameterLikelihood(iio.YAMLLoadable):
+    """An `UnbinnedParameterLikelihood` provides an unbinned likelihood function in terms of
+    parameters.
+    Methods:
+    - `log_likelihood`: The likelihood as a function of the parameters
+    Instances can be imported and exported from/to YAML using the `load`
+    and `dump` methods.
+    """
+    _input_schema_dict = {
+        'eft': str,
+        'basis': str,
+        'scale': float,
+        'names': list,
+        'samples': str,
+        'weights': str
+    }
+
+    _output_schema_dict = {
+        'eft': str,
+        'basis': str,
+        'scale': float,
+        'names': list,
+        'samples': str,
+        'weights': str
+    }
+
+    def __init__(self,
+                 eft,
+                 basis,
+                 scale,
+                 names,
+                 samples,
+                 base_path,
+                 weights=None):
+        """Initialize the instance.
+        Parameters:
+        - `eft`: a string identifying the WET as defined by wcxf
+        - `basis`: a string identifying the WET basis as defined by wcxf
+        - `scale`: a scale (in units of GeV) at which the WET coefficient samples are valid
+        - `names`: an ordered list of names identiying the Wilson coefficients in the WET basis
+        - `samples`: a numpy array of vectors of Wilson coefficients, with each row representing a sample
+        - `weights`: a numpy array of scalar weights on a logarithmic scale for each sample
+        """
+        self.eft     = eft
+        self.basis   = basis
+        self.scale   = scale
+        self.names   = names
+        self.samples = np.load(os.path.join(base_path, samples))
+        self.weights = np.ones(len(self.samples)) if weights is None else np.load(os.path.join(base_path, weights))
+        self.kde     = KernelDensity(algorithm='auto', kernel='gaussian', rtol=1e-4)
+
+        basis = wcxf.Basis[eft, basis]
+        basis_wcs = {}
+        for sector in basis.sectors.values():
+            basis_wcs.update(sector)
+
+        self.accessors = []
+        for i, n in enumerate(names):
+            if 'real' in basis_wcs[n]:
+                self.accessors.append(lambda wcs: wcs[i])
+            else:
+                self.accessors.append(lambda wcs: np.real(wcs[i]))
+                self.accessors.append(lambda wcs: np.imag(wcs[i]))
```

Comment:
@dvandyk I think this doesn't work. `i` should be overwritten in each iteration such that in the end, `i` will be just `len(names)-1`. So I think all functions in `self.accessors` would use this final `i` when they are evaluated in the following and thus all values in each row will be equal to the value of the last Wilson coefficient of that row.

Reply:
Another issue might be that `if 'real' in basis_wcs[n]`, `wcs[i]` is just copied without turning it to a real number. So if it stems from a numpy array of complex type, it's imaginary part might just be `1e-16` but it would still be of complex type. This then will make any numpy array constructed with this number as an entry complex again. I don't know if this could be a problem for the kde. So I would also use `np.real()` in this case just to be sure that no complex numbers remain.

Reply:
> Another issue might be that `if 'real' in basis_wcs[n]`, `wcs[i]` is just copied without turning it to a real number. ...
Acknowledged and changed locally. Thanks Peter!

Reply:
> @dvandyk I think this doesn't work. `i` should be overwritten in each iteration such that in the end, `i` will be just `len(names)-1`. So I think all functions in `self.accessors` would use this final `i` when they are evaluated in the following and thus all values in each row will be equal to the value of the last Wilson coefficient of that row.

Peter, thank you very much! I tested and confirmed it at the hand of this snippet:
```
sin = ['m', 'l', 'i', 'l', 'e', 's']
idx = [5, 0, 4, 1, 3, 2]
acc = []
for i in idx:
    acc.append(lambda mys: mys[i])

sout = [a(sin) for a in acc]
sout
```
The output is indeed ```['i', 'i', 'i', 'i', 'i', 'i']```. This is completely unexpected for me,
and does not follow the idea of a lambda functional at all, but that's a Python issue.

It works if I return the lambda from a different function, though. Using that for the time being.